File: Java/121.Best_Time_to_Buy_and_Sell_Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length <= 1) return 0;

        int cur_min = prices[0], cur_max = prices[0], profit = 0;

        for (int price : prices) {
            if (price < cur_min) {
                cur_min = cur_max = price;
            }
            else if (price > cur_max) {
                cur_max = price;
                profit = Math.max(profit, cur_max - cur_min);
            }
        }

        return profit;
    }
}

File: Python3/121. Best_Time_to_Buy_and_Sell_Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) <= 1:
            return 0
        
        cur_min = cur_max = prices[0]
        cur_profit = 0

        for price in prices:
            if price < cur_min:
                cur_min = cur_max = price
            if price > cur_max:
                cur_max = price
                cur_profit = max(cur_profit, cur_max - cur_min)

        
        return cur_profit

                 

